File: php-conpackist/tanstackship-utm-parse/src/UtmParser.php
<?php

declare(strict_types=1);

namespace TanStackShip\UtmParse;

/**
 * UTM URL Parser for PHP
 */
final class UtmParser
{
    private const PLATFORM_CLICK_IDS = [
        'gclid' => 'google',
        'fbclid' => 'facebook',
        'msclkid' => 'bing',
        'ttclid' => 'tiktok',
        'li_fat_id' => 'linkedin',
    ];

    private const NON_STANDARD_SOURCE_PARAMS = [
        'ref',
        'referrer',
        'refer',
        'source',
        'via',
        'from',
        'origin',
        'src',
    ];

    /**
     * Parse UTM parameters from URL string
     *
     * @param string|null $urlString URL to parse
     * @return Params UTM params
     *
     * @example
     * ```php
     * $params = UtmParser::parse("https://example.com?utm_source=google");
     * echo $params->source; // "google"
     * ```
     */
    public static function parse(?string $urlString): Params
    {
        $params = new Params();

        if ($urlString === null || $urlString === '') {
            return $params;
        }

        $parsed = parse_url($urlString);

        if (!isset($parsed['query'])) {
            return $params;
        }

        parse_str($parsed['query'], $query);

        // Parse standard UTM params
        $params->source = $query['utm_source'] ?? null;
        $params->medium = $query['utm_medium'] ?? null;
        $params->campaign = $query['utm_campaign'] ?? null;
        $params->term = $query['utm_term'] ?? null;
        $params->content = $query['utm_content'] ?? null;

        // If no utm_source, try platform click IDs
        if ($params->source === null) {
            foreach (self::PLATFORM_CLICK_IDS as $clickId => $platform) {
                if (isset($query[$clickId])) {
                    $params->source = $platform;
                    $params->platform = $platform;
                    break;
                }
            }
        }

        // If still no source, try non-standard params
        if ($params->source === null) {
            foreach (self::NON_STANDARD_SOURCE_PARAMS as $param) {
                if (isset($query[$param])) {
                    $params->source = $query[$param];
                    break;
                }
            }
        }

        return $params;
    }
}

File: .net/TanStackShip.UtmParse/UtmParser.cs
using System.Web;

namespace TanStackShip.UtmParse;

/// <summary>
/// UTM URL Parser for .NET
/// </summary>
public static class UtmParser
{
    private static readonly Dictionary<string, string> PlatformClickIds = new()
    {
        ["gclid"] = "google",
        ["fbclid"] = "facebook",
        ["msclkid"] = "bing",
        ["ttclid"] = "tiktok",
        ["li_fat_id"] = "linkedin"
    };

    private static readonly string[] NonStandardSourceParams = 
    {
        "ref", "referrer", "refer", "source", "via", "from", "origin", "src"
    };

    /// <summary>
    /// Parse UTM parameters from URL string
    /// </summary>
    /// <param name="urlString">URL to parse</param>
    /// <returns>UtmParams object</returns>
    /// <example>
    /// ```csharp
    /// var params = UtmParser.Parse("https://example.com?utm_source=google");
    /// Console.WriteLine(params.Source); // "google"
    /// ```
    /// </example>
    public static UtmParams Parse(string urlString)
    {
        var result = new UtmParams();

        if (string.IsNullOrEmpty(urlString))
            return result;

        try
        {
            var uri = new Uri(urlString);
            var query = HttpUtility.ParseQueryString(uri.Query);

            // Parse standard UTM params
            result.Source = query["utm_source"];
            result.Medium = query["utm_medium"];
            result.Campaign = query["utm_campaign"];
            result.Term = query["utm_term"];
            result.Content = query["utm_content"];

            // If no utm_source, try platform click IDs
            if (string.IsNullOrEmpty(result.Source))
            {
                foreach (var (param, platform) in PlatformClickIds)
                {
                    if (!string.IsNullOrEmpty(query[param]))
                    {
                        result.Source = platform;
                        result.Platform = platform;
                        break;
                    }
                }
            }

            // If still no source, try non-standard params
            if (string.IsNullOrEmpty(result.Source))
            {
                foreach (var param in NonStandardSourceParams)
                {
                    if (!string.IsNullOrEmpty(query[param]))
                    {
                        result.Source = query[param];
                        break;
                    }
                }
            }
        }
        catch
        {
            // Invalid URL, return empty params
        }

        return result;
    }
}
